application/app/controllers/upload_files_controller.rb

```diff
@@ -40,69 +32,91 @@ def create
   end
 
   def destroy
-    project_id = params[:project_id]
-    upload_bundle_id = params[:upload_bundle_id]
-    file_id = params[:id]
-    upload_file = UploadFile.find(project_id, upload_bundle_id, file_id)
+    if @upload_file.status.uploading?
+      return redirect_back fallback_location: root_path, alert: t(".file_in_progress", filename: @upload_file.filename)
+    end
 
-    if upload_file.nil?
-      redirect_back fallback_location: root_path, alert: t(".file_not_found", file_id: file_id, project_id: project_id)
-      return
+    @upload_file.destroy
+    log_upload_file_event(@upload_file, message: 'events.upload_file.deleted')
+    log_info('Upload file deleted', {id: @upload_file.id, filename: @upload_file.filename})
+    redirect_back fallback_location: root_path, notice: t(".upload_file_removed", filename: @upload_file.filename)
+  end
+
+  def cancel
+    previous_status = @upload_file.status.to_s
+    if @upload_file.status.uploading?
+      command_client = Command::CommandClient.new(socket_path: ::Configuration.command_server_socket_file)
+      request = Command::Request.new(command: 'upload.cancel', body: { project_id: @upload_file.project_id, upload_bundle_id: @upload_file.upload_bundle_id, file_id: @upload_file.id})
+      response = command_client.request(request)
+      return redirect_back fallback_location: root_path, alert: t('.file_cancellation_error', filename: @upload_file.filename) if response.status != 200
     end
 
-    if upload_file.status.uploading?
-      redirect_back fallback_location: root_path, alert: t(".file_in_progress", filename: upload_file.filename)
-      return
+    if @upload_file.update(status: FileStatus::CANCELLED)
+      log_upload_file_event(@upload_file, message: 'events.upload_file.cancel_completed', metadata: { previous_status: previous_status })
+      log_info('Upload file cancelled', {id: @upload_file.id, filename: @upload_file.filename})
+      redirect_back fallback_location: root_path, notice: t('.file_cancellation_success', filename: @upload_file.filename)
+    else
+      redirect_back fallback_location: root_path, notice: t('.file_cancellation_update_error', filename: @upload_file.filename)
     end
+  end
 
-    upload_file.destroy
-    redirect_back fallback_location: root_path, notice: t(".upload_file_removed", filename: upload_file.filename)
+  def retry
+    previous_status = @upload_file.status
+    if @upload_file.update(status: FileStatus::PENDING)
+      log_upload_file_event(@upload_file, message: 'events.upload_file.retry_request', metadata: { previous_status: previous_status })
+      log_info('Upload file retry requested', {id: @upload_file.id, filename: @upload_file.filename})
+      redirect_back fallback_location: root_path, notice: t('.file_retry_success', filename: @upload_file.filename)
+    else
+      redirect_back fallback_location: root_path, alert: t('.file_retry_error', filename: @upload_file.filename)
+    end
   end
```

Comment:
I would check on the server side that the file is retrievable, since it seems this is only checked on the view. I guess that checking that it is in Cancelled or Error status is enough. I would do the same on the DownloadFilesController